src/System.Linq.Expressions/tests/BinaryOperators/Logical/BinaryLogicalTests.cs

```diff
@@ -142,15 +145,180 @@ public static void CannotReduceOrElse()
+            Assert.Throws<ArgumentException>(null, () => Expression.AndAlso(Expression.Constant(5), Expression.Constant(5), method));
+        }
+
+        public static IEnumerable<object[]> Operator_IncorrectMethod_TestData()
+        {
+            // Does not return bool
+            TypeBuilder typeBuilder1 = GetTypeBuilder();
+            yield return new object[] { typeBuilder1, typeof(void), new Type[] { typeBuilder1.AsType() } };
+            
+            // Parameter is not assignable from left
+            yield return new object[] { GetTypeBuilder(), typeof(bool), new Type[] { typeof(int) } };
+
+            // Has two parameters
+            TypeBuilder typeBuilder2 = GetTypeBuilder();
+            yield return new object[] { typeBuilder2, typeof(bool), new Type[] { typeBuilder2.AsType(), typeBuilder2.AsType() } };
+
+            // Has no parameters
+            yield return new object[] { GetTypeBuilder(), typeof(bool), new Type[0] };
+        }
+
+        [Theory]
+        [MemberData(nameof(Operator_IncorrectMethod_TestData))]
+        public static void AndAlso_TrueOperatorIncorrectMethod_ThrowsArgumentException(TypeBuilder builder, Type returnType, Type[] parameterTypes)
+        {
+            MethodBuilder opTrue = builder.DefineMethod("op_True", MethodAttributes.SpecialName | MethodAttributes.Static, returnType, parameterTypes);
+            opTrue.GetILGenerator().Emit(OpCodes.Ret);
+
+            MethodBuilder opFalse = builder.DefineMethod("op_False", MethodAttributes.SpecialName | MethodAttributes.Static, typeof(bool), new Type[] { builder.AsType() });
+            opFalse.GetILGenerator().Emit(OpCodes.Ret);
+
+            MethodBuilder method = builder.DefineMethod("Method", MethodAttributes.Public | MethodAttributes.Static, builder.AsType(), new Type[] { builder.AsType(), builder.AsType() });
+            method.GetILGenerator().Emit(OpCodes.Ret);
+
+            TypeInfo createdType = builder.CreateTypeInfo();
+            object obj = Activator.CreateInstance(createdType.AsType());
+            MethodInfo createdMethod = createdType.GetMethod("Method");
+
+            Assert.Throws<ArgumentException>(null, () => Expression.AndAlso(Expression.Constant(obj), Expression.Constant(obj), createdMethod));
+        }
+
+        [Theory]
+        [MemberData(nameof(Operator_IncorrectMethod_TestData))]
+        public static void AndAlso_FalseOperatorIncorrectMethod_ThrowsArgumentException(TypeBuilder builder, Type returnType, Type[] parameterTypes)
+        {
+            MethodBuilder opTrue = builder.DefineMethod("op_True", MethodAttributes.SpecialName | MethodAttributes.Static, typeof(bool), new Type[] { builder.AsType() });
+            opTrue.GetILGenerator().Emit(OpCodes.Ret);
+
+            MethodBuilder opFalse = builder.DefineMethod("op_False", MethodAttributes.SpecialName | MethodAttributes.Static, returnType, parameterTypes);
+            opFalse.GetILGenerator().Emit(OpCodes.Ret);
```

Comment:
... and the reason why I had to use ref-emit was that C# doesn't allow for this invalid code for the true or false code. Other compilers may do, but anyway product code makes sure we don't allow any invalid inputs. It's easier to create using ref-emit than IL